pkg/controller/autoops/expected.go

```diff
@@ -0,0 +1,265 @@
+	configHashAnnotationName = "autoops.k8s.elastic.co/config-hash"
+	readinessProbePort       = 13133
+)
+
+var (
+	// ESNAutoOpsNamer is a Namer that generates names for AutoOps deployments
+	// according to the Policy name, and associated Elasticsearch name.
+	AutoOpsNamer    = common_name.NewNamer("autoops")
+	basePodTemplate = corev1.PodTemplateSpec{
+		Spec: corev1.PodSpec{
+			Containers: []corev1.Container{
+				{
+					Name: "autoops-agent",
+				},
+			},
+		},
+	}
+)
+
+type ExpectedResources struct {
+	deployment appsv1.Deployment
+}
+
+func (r *ReconcileAutoOpsAgentPolicy) generateExpectedResources(ctx context.Context, policy autoopsv1alpha1.AutoOpsAgentPolicy, es esv1.Elasticsearch) (ExpectedResources, error) {
+	deployment, err := r.deploymentParams(ctx, policy, es)
+	if err != nil {
+		return ExpectedResources{}, err
+	}
+	return ExpectedResources{
+		deployment: deployment,
+	}, nil
+}
+
+func (r *ReconcileAutoOpsAgentPolicy) deploymentParams(ctx context.Context, policy autoopsv1alpha1.AutoOpsAgentPolicy, es esv1.Elasticsearch) (appsv1.Deployment, error) {
+	var deployment appsv1.Deployment
+	v, err := version.Parse(policy.Spec.Version)
+	if err != nil {
+		return appsv1.Deployment{}, err
+	}
+	labels := map[string]string{
+		commonv1.TypeLabelName: "autoops-agent",
+		autoOpsLabelName:       policy.Name,
+	}
+	// Hash ES namespace and name to create a short unique identifier
+	// preventing name length issues.
+	esIdentifier := es.GetNamespace() + es.GetName()
+	esHash := fmt.Sprintf("%x", sha256.Sum256([]byte(esIdentifier)))[0:6]
+	deployment.ObjectMeta = metav1.ObjectMeta{
+		Name:      AutoOpsNamer.Suffix(policy.GetName(), esHash),
+		Namespace: policy.GetNamespace(),
+		Labels:    labels,
+	}
+	deployment.Spec = appsv1.DeploymentSpec{
+		Replicas: pointer.Int32(1),
+		Selector: &metav1.LabelSelector{
+			MatchLabels: map[string]string{
+				autoOpsLabelName: policy.GetName(),
+			},
+		},
+	}
+	// Create ES-specific config map volume
+	configMapName := fmt.Sprintf("%s-%s-%s", autoOpsESConfigMapName, es.Namespace, es.Name)
+	configVolume := volume.NewConfigMapVolume(configMapName, configVolumeName, configVolumePath)
+
+	volumes := []corev1.Volume{configVolume.Volume()}
+	volumeMounts := []corev1.VolumeMount{configVolume.VolumeMount()}
+
+	// Add CA certificate volume for this ES instance only if TLS is enabled
+	if es.Spec.HTTP.TLS.Enabled() {
+		caSecretName := fmt.Sprintf("%s-%s-%s", es.Name, es.Namespace, autoOpsESCASecretSuffix)
+		caVolume := volume.NewSecretVolumeWithMountPath(
+			caSecretName,
+			fmt.Sprintf("es-ca-%s-%s", es.Name, es.Namespace),
+			fmt.Sprintf("/mnt/elastic-internal/es-ca/%s-%s", es.Namespace, es.Name),
+		)
+		volumes = append(volumes, caVolume.Volume())
+		volumeMounts = append(volumeMounts, caVolume.VolumeMount())
+	}
+
+	// Build config hash from ConfigMap to trigger pod restart on config changes
+	configHash, err := buildConfigHash(ctx, r.Client, policy, es)
```

Comment:
since in the config we have env vars that derive from a secret I do feel that this hash should take the latter into consideration and restart the pod

Reply:
I've also added the hashing of the secret data that's used in these env vars, and updated the unit tests.